File: HeapSort/HeapSort.js
const sort = (inputList) => {
    inputList = buildMaxHeap(inputList);

    for(let i = inputList.length - 1; i >= 0; i--) {
        [inputList[0], inputList[i]] = [inputList[i], inputList[0]];
        inputList = maxHeapify(inputList, 0, i);
    }

    return inputList;
}

const buildMaxHeap = (inputList) => {
    for( let i = Math.round(inputList.length / 2); i >= 0; i--) {
             
        maxHeapify(inputList, i);
    }

    return inputList;
}

const maxHeapify = (inputList, rootIndex, length=inputList.length) => {
    let maxIndex = rootIndex;    
    const leftChildIndex = getLeftChildIndex(rootIndex);
    const rightChildIndex = getRightChildIndex(rootIndex);

    if( (leftChildIndex < length) && (inputList[maxIndex] < inputList[leftChildIndex]) ) {
        maxIndex = leftChildIndex;
    }

    if( (rightChildIndex < length) && (inputList[maxIndex] < inputList[rightChildIndex]) ) {
        maxIndex = rightChildIndex;
    }

    if(maxIndex !== rootIndex) {
        [inputList[rootIndex], inputList[maxIndex]] = [inputList[maxIndex], inputList[rootIndex]];
        maxHeapify(inputList, maxIndex, length);
    }

    return inputList;
}

const getLeftChildIndex = parentIndex => parentIndex * 2 + 1;

const getRightChildIndex = parentIndex => parentIndex * 2 + 2;


console.log(`sort(2, 6, 3, 9, 1, 5) => ${sort([2, 6, 3, 9, 1, 5])}`);
console.log(`sort(2, 6, 3, 9, 1) => ${sort([2, 6, 3, 9, 1])}`);

File: HeapSort/HeapSort.go
package main

import "fmt"

func main() {
	fmt.Println("sort(2, 6, 3, 9, 1, 5) => ", sort([]int{2, 6, 3, 9, 1, 5}))
	fmt.Println("sort(2, 6, 3, 9, 1) => ", sort([]int{2, 6, 3, 9, 1}))
}

func sort(inputList []int) []int {
	inputList = buildMaxHeap(inputList)

	for i:= len(inputList) - 1; i >= 0; i-- {
		inputList[i], inputList[0] = inputList[0], inputList[i]
		maxHeapify(inputList[0:i], 0)
	}

	return inputList
}

func buildMaxHeap(inputList []int) []int {
	for i := len(inputList)  / 2; i >= 0; i-- {
		inputList = maxHeapify(inputList, i)
	}

	return inputList;
}

func maxHeapify(inputList []int, rootIndex int) []int {
	var maxIndex int = rootIndex
	var leftChildIndex int = getLeftChildIndex(rootIndex)
	var rightChildIndex int = getRightChildIndex(rootIndex)

	if (leftChildIndex < len(inputList)) && (inputList[maxIndex] < inputList[leftChildIndex]) {
		maxIndex = leftChildIndex
	}

	if (rightChildIndex < len(inputList)) && (inputList[maxIndex] < inputList[rightChildIndex]) {
		maxIndex = rightChildIndex
	}

	if maxIndex != rootIndex {
		inputList[maxIndex], inputList[rootIndex] = inputList[rootIndex], inputList[maxIndex]
		return maxHeapify(inputList, maxIndex)
	}

	return inputList

}

func getLeftChildIndex(parentIndex int) int {
	return 2 * parentIndex + 1
}

func getRightChildIndex(parentIndex int) int {
	return 2 * parentIndex + 2
}